### PR TITLE
Ad/4211 MMIS APD overview

### LINKED ISSUE #4211: [Dev] Implement MMIS APD Overview
### Description and related issues
The MMIS APD Overview page will be different from the current APD Overview page.  The fields from the Create new APD page will live in this page. **APD Type cannot be changed**

**Update the component to have the following shared fields:**
* APD Name
* What type of APD are you creating (read-only, disabled)
* FFY 
* Is an APD Update?

**for HITECH the fields will stay the same:**
* Introduction
* HIT Overview
* HIE Overview
* MMIS Overview

**for MMIS the fields will be:**
* Medicaid Business Area

### Behind the scene changes:
* adding new set methods in actions/editApd/overview.js to handle saving the new fields

Schema for updated HIT:
```json
...
apdOverview: {
  updateStatus: {
    isUpdateAPD: Boolean,
    annualUpdate: Boolean,
    asNeededUpdate: Boolean
  },
  programOverview: String,
  narrativeHIT: String,
  narrativeHIE: String,
  narrativeMMIS: String
},
...
```

Schema for update MMIS:
```json
...
apdOverview: {
  updateStatus: {
     isUpdateAPD: Boolean,
     annualUpdate: Boolean,
     asNeededUpdate: Boolean
   },
   medicaidBusinessAreas: {
     waiverSupportSystems: Boolean,
     assetVerificationSystem: Boolean,
     claimsProcessing: Boolean,
     decisionSupportSystemDW: Boolean,
     electronicVisitVerification: Boolean,
     encounterProcessingSystemMCS: Boolean,
     financialManagement: Boolean,
     healthInformationExchange: Boolean,
     longTermServicesSupports: Boolean,
     memberManagement: Boolean,
     pharmacyBenefitManagementPOS: Boolean,
     programIntegrity: Boolean,
     providerManagement: Boolean,
     thirdPartyLiability: Boolean,
     other: Boolean,
     otherMedicaidBusinessAreas: String
   }
},
...
```

**Update the error message for APD Name**
The current behavior of the APD name field does not allow the user to enter a blank name. On blur, if the field is blank, the name will default to "Untitled IAPD". So for validation on this field, check if the name matches this default name.

Validation error message language: "Provide your own APD Name"

Design ticket: #4179
Figma link:  https://www.figma.com/file/hJpKHKU6fz5J0Z7fisSwa2/eAPD-MMIS-2022?node-id=414%3A10804

Test Cases:
| Given  | When | Then | 
| ------------- | ------------- | ------------- |
| APD Type Field | On the APD Overview page | This field cannot be changed and should be grayed out |
| An invalid APD Name | blurring out of the field | the error message will appear whether adminCheck is on or off |
| APD Overview Page | Admin check is on | If field is blank, inline validation should appear. Refer to this dev ticket for error language: #4209
| APD Overview Page | Right after creating a new APD | Green header should appear on the page informing the user they successfully created an APD. Header should **only** appear after first creating an APD.| 
| APD Overview Page | Right after creating a new APD | Selections should be the same from the Create new APD page | 
| APD Overview Page | Clicking the Continue button | Should redirect you to the state scope and priorities page (out of scope of this ticket) | 

### This task is done when...

- [ ] APD Overview for MMIS has been updated to match the designs
- [ ] APD Overview for HITECH still works and looks like previously did, but with the new shared field
- [ ] validation is working for APD Name on blur
- [ ] validation is working for the rest of the form when admin check is on
- [ ]  admin check catches when an APD has the name "Untitled"
- [ ] APD name field in create new APD page has inline validation on blur when name defaults to "Untitled"

### PR DESCRIPTION
**Pull request by amyd11**
Resolves #4211 

### Description
- Updating APD Overview to allow an MMIS APD type. This includes removing fields specific to HITECH and adding a field specific to MMIS
- Addition of APD Update field to APD Overview page
- Adds validation to APD Name

### Significant changes or possible side effects
- Schema updates:
  - MMIS medicaidBusinessAreas tweaking for validation purposes
  - APD Name schema includes ensuring the name does not contain "untitled"
  - APD Update schema created for each of MMIS and HITECH since the validations differ
- Redux action additions:
  - action to update a key value pair in the `medicaidBusinessAreas` object
  - action to update a key value pair in the `updateStatus` object
- Component changes:
  - Set up APD Name in `ApdOverview` to be handled within react hook form for validation purposes
  - Modified `ApdOverview` to contain only the shared fields of HITECH and MMIS APDs. Subsequently moved HITECH fields to new component called `ApdOverviewHITECHFields` and created `ApdOverviewMMISFields` for the MMIS fields.
  - Created `ApdUpdate` component to handle displaying the Update APD section differently for MMIS and HITECH types
  - Added `disableTitleCaseConversion` prop to the `Instruction` component
- Backend
  -  Added APD Name validation to fall under APD Overview in Admin Check
- Misc
  - Created a constant for mapping Medicaid Business Areas key names to labels
  - Updated seed data for invalid and valid options regarding updateStatus field
  - Updated label sizing for fields on APD Overview page
  - Resolved some existing console warnings for "APD Type" field by adding `name` and `label` props. Since the code was written with the intent of the label being hidden, a `labelClassName` prop was also passed in to include a newly added `hidden-display` style class which hides the label

### Automated test cases written

| Given | When | Then | Type (jest, tap, cypress) |
| ----- | ---- | ---- | ------------------------- |
APD Name with Admin Check ON | Name contains "untitled" | Validation error appears on the field and in admin check | Cypress
APD Name | Previous data exists for name | Displays previous data in the APD Name field | Jest
MMIS APD Overview medicaid business areas  | No option selected | Validation error appears on the field and in admin check | Cypress
MMIS APD Overview medicaid business areas  | "Other" option selected, but "Other Medicaid Business Areas" text box is empty | Validation error appears on the field and in admin check | Cypress
MMIS APD Overview medicaid business areas with Admin Check ON | Form errors are active | Validation error also appears in admin check | Cypress
MMIS Medicaid Business Areas | Previous data exists | Displays checkbox as selected or unselected based on existing data | Jest
MMIS Medicaid Business Areas | "Other" option is checked | "Other Medicaid Business Area(s)" text area and content displays | Jest
MMIS Medicaid Business Areas | "Other" option is NOT checked | "Other Medicaid Business Area(s)" text area is NOT visible | Jest
HITECH APD Overview | Each of the text fields | Allows text changes | Jest
APD Update in APD Overview | HITECH type | Displays "Update Type" checkbox options | Jest & Cypress
APD Update in APD Overview | MMIS type | Displays "Is this an APD update?" radio options | Jest & Cypress
MMIS APD Overview APD Update field | "Yes, it is an update." radio option is selected | "Update Type" options available | Jest & Cypress
APD Overview APD Update field with Admin Check ON | "Update Type" options are all unchecked | Validation error appears on the field and in admin check | Cypress
Instruction component | default behavior for heading | heading is converted to title case | Jest
Instruction component | `disableTitleCaseConversion` is passed in | original heading text appears as is (NOT converted to title case) | Jest
APD edit actions for Medicaid Business Areas | `setBusinessAreaField` is called | redux dispatches an action for changing the field value | unit test

### Steps to verify changes
_With admin check ON_
1. Open an existing MMIS APD
1. Go to APD Overview
2. Test APD Name
    1. Modify the APD name (try including "untitled" in the name) to ensure that errors display both on the form AND in admin check
    4. Modify the APD name _from the APD Header_ and ensure that the name changes reflect within the APD Overview form and that validation errors accurately change on the form.
5. Test APD Update 
    1. Select "Yes, it is an update."
    2. Ensure that the "Update Type" section only appears when "Yes, it is an update." is selected.
    7. Ensure error to "Select an update type." appears in admin check and on the form.
    8. Ensure error disappears when either an update type is selected or the radio button choice is switched to "No, this is for a new project."
1. Test Medicaid Business Areas
    1. Unselect any checked Medicaid Business Areas to ensure the validation error requiring a selection display both on the form AND in admin check
    11. Select "Other" checkbox in Medicaid Business Areas to ensure a text box appears for "Other Medicaid Business Areas". Check that validation error for text box displays when empty (both on the form AND in admin check)

_With admin check OFF_
1. Go to an existing MMIS APD
3. Ensure that none of the validation errors appear on the form

_Other testing_
1. HITECH APD Overview
    1. Regression testing
    3. Update Type field displays
    4. _With admin check on,_ ensure error to "Select an update type." appears in admin check and on the form

### This pull request is ready to code review when

- [x] Automated tests are updated (and all tests are passing)
- [x] New automated test cases are documented above
- [x] Pull request has been labeled, if applicable with feature, content, bug,
      tests, refactor
- [ ] The experience passes a basic manual accessibility audit (keyboard nav,
      screenreader, text scaling) OR an exemption is documented

### This pull request is ready to test when

- [x] Code has been reviewed by someone other than the original author

### This pull request is ready to review when the QA has

- [ ] Verified the functionality related to the change
- [ ] Verified that the change works with Narrator on Windows
- [ ] Verified that the change works with VoiceOver on Mac
- [ ] Verified all updated pages with the WAVE tool
- [ ] Verified tab and keyboard navigation functionality

### This pull request can be merged when

- [ ] Design has approved the experience
- [ ] Product has approved the experience

### CODE REVIEW

**Review by thetif — APPROVED** [review]

File: web/src/pages/apd/apd-overview/ApdOverview.js [hunk 1]
```diff
@@ -231,13 +202,24 @@ const ApdOverview = ({
...
-        onChange={changeName}
-        onBlur={onBlur}
-        value={name}
+      <Controller
+        name="name"
+        control={control}
+        render={() => {
+          return (
+            <TextField
+              {...register('name')}
```
[the remark this answers is not in the document]
Comment (amyd11) [reply web/src/pages/apd/apd-overview/ApdOverview.js]: I totally forgot to take this out once I wrapped it with the Controller component. Thanks for noticing it! Removed ✅

File: api/util/adminCheck.test.js [hunk 2]
```diff
@@ -2135,7 +2135,7 @@ const expectedErrorsMockMmisApdRandom = [
   {
     section: 'APD Overview',
     link: '/apd/632a0fbc5665670a34b3bbd7/apd-overview',
-    fieldDescription: 'Provide another Medicaid Business Area(s)'
+    fieldDescription: 'Provide an other Medicaid Business Area(s).'
```
[the remark this answers is not in the document]
Comment (amyd11) [reply api/util/adminCheck.test.js]: **For record:** Per standup the standup convo today, we are going to use the message "Provide Other Medicaid Business Area(s)."

File: web/src/pages/apd/apd-overview/ApdOverview.js [hunk 3]
```diff
@@ -35,26 +27,28 @@ import {
...
   selectAdminCheckEnabled
 } from '../../../redux/selectors/apd.selectors';
-import { getAllFundingSources } from '../../../redux/selectors/activities.selectors';
 import { useFlags } from 'launchdarkly-react-client-sdk';
 
+import ApdOverviewHITECHFields from './ApdOverviewHITECHFields';
+import ApdOverviewMMISFields from './ApdOverviewMMISFields';
+import DeleteModal from '../../../components/DeleteModal';
+
+const renderApdTypeSpecificFields = apdType => {
```
[the remark this answers is not in the document]
Comment (tbolt) [reply web/src/pages/apd/apd-overview/ApdOverview.js:36]: agreed. this is cleaner than the ternary way

**Review by tbolt — APPROVED** [review]
Nice work

**Review by Sun-Mountain — APPROVED** [review]

File: common/schemas/apdOverview.js [hunk 4]
```diff
@@ -1,29 +1,59 @@
...
-  asNeededUpdate: Joi.when('isUpdateAPD', {
-    is: true,
-    then: Joi.boolean(),
-    otherwise: Joi.any()
-  })
+  annualUpdate: Joi.boolean(),
+  asNeededUpdate: Joi.boolean()
 })
-  .or('annualUpdate', 'asNeededUpdate')
+  .custom(updateStatusCustomValidation)
```
Comment (thetif) [inline common/schemas/apdOverview.js:22]: 🧠

**Review by thetif — APPROVED** [review]

**Comment by mirano-darren** [comment]
Looks good to me! Nice work with the APD Name validation 😄 
Test Cases:
| Given  | When | Then | Result | 
| ------------- | ------------- | ------------- | ------------- | 
| APD Type Field | On the APD Overview page | This field cannot be changed and should be grayed out | ✅ |
| An invalid APD Name | blurring out of the field | the error message will appear whether adminCheck is on or off |✅ |
| APD Overview Page | Admin check is on | If field is blank, inline validation should appear. Refer to this dev ticket for error language: #4209|✅ |
| APD Overview Page | Right after creating a new APD | Green header should appear on the page informing the user they successfully created an APD. Header should **only** appear after first creating an APD.| ✅ |
| APD Overview Page | Right after creating a new APD | Selections should be the same from the Create new APD page | ✅ |
| APD Overview Page | Clicking the Continue button | Should redirect you to the state scope and priorities page (out of scope of this ticket) |✅ |

**Review by mirano-darren — APPROVED** [review]

**Review by SGilliamA1M — APPROVED** [review]
Changes look good

**Pull request merged**